File: src/sandbox/docker.ts
import { Effect, Layer } from "effect";
import { randomUUID } from "node:crypto";
import { execFileSync, spawn } from "node:child_process";
import * as fs from "node:fs";
import {
  ContainerBackendTag,
  type ContainerBackend,
} from "./backend.js";
import { ContainerError, type ContainerRunResult } from "./schema.js";

const DEFAULT_TIMEOUT_MS = 120_000;

const hasDockerCommand = (): boolean => {
  try {
    if (process.env.OPENAGENTS_DOCKER_AVAILABLE === "0") return false;
    if (process.env.OPENAGENTS_DOCKER_AVAILABLE === "1") return true;
    execFileSync("docker", ["--version"], { stdio: "ignore", timeout: 2000 });
    return true;
  } catch {
    return false;
  }
};

const buildDockerArgs = (
  command: string[],
  config: {
    image: string;
    workspaceDir: string;
    workdir?: string;
    memoryLimit?: string;
    cpuLimit?: number;
    env?: Record<string, string>;
    volumeMounts?: string[];
    autoRemove?: boolean;
  },
  containerName: string,
) => {
  const args: string[] = ["run", "-i", "--name", containerName];

  if (config.autoRemove ?? true) {
    args.push("--rm");
  }

  args.push("-v", `${config.workspaceDir}:/workspace`);

  for (const mount of config.volumeMounts ?? []) {
    args.push("-v", mount);
  }

  args.push("-w", config.workdir ?? "/workspace");

  if (config.memoryLimit) {
    args.push("--memory", config.memoryLimit);
  }
  if (config.cpuLimit) {
    args.push("--cpus", String(config.cpuLimit));
  }

  for (const [key, value] of Object.entries(config.env ?? {})) {
    args.push("-e", `${key}=${value}`);
  }

  args.push(config.image);
  args.push(...command);

  return args;
};

const dockerRun: ContainerBackend["run"] = (command, config, options) =>
  Effect.async<ContainerRunResult, ContainerError>((resume) => {
    if (!fs.existsSync(config.workspaceDir)) {
      resume(
        Effect.fail(
          new ContainerError(
            "start_failed",
            `workspaceDir does not exist: ${config.workspaceDir}`,
          ),
        ),
      );
      return;
    }

    const name = `oa-sbx-${randomUUID()}`;
    const runConfig = {
      image: config.image,
      workspaceDir: config.workspaceDir,
      ...(config.workdir ? { workdir: config.workdir } : {}),
      ...(config.memoryLimit ? { memoryLimit: config.memoryLimit } : {}),
      ...(config.cpuLimit !== undefined ? { cpuLimit: config.cpuLimit } : {}),
      ...(config.env ? { env: config.env } : {}),
      ...(config.volumeMounts ? { volumeMounts: [...config.volumeMounts] } : {}),
      ...(config.autoRemove !== undefined ? { autoRemove: config.autoRemove } : {}),
    };
    const args = buildDockerArgs(command, runConfig, name);
    const proc = spawn("docker", args, { stdio: ["ignore", "pipe", "pipe"] });

    let stdout = "";
    let stderr = "";
    let finished = false;

    const cleanup = () => {
      finished = true;
      clearTimeout(timeoutHandle);
    };

    const finish = (result: Effect.Effect<ContainerRunResult, ContainerError>) => {
      cleanup();
      resume(result);
    };

    const timeoutMs = config.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    const timeoutHandle = setTimeout(() => {
      if (finished) return;
      proc.kill("SIGKILL");
      finish(
        Effect.fail(
          new ContainerError(
            "timeout",
            `Container execution timed out after ${timeoutMs}ms`,
          ),
        ),
      );
    }, timeoutMs);

    proc.on("error", (error) => {
      if (finished) return;
      finish(
        Effect.fail(
          new ContainerError(
            "start_failed",
            `Failed to start docker: ${error.message}`,
          ),
        ),
      );
    });

    proc.stdout?.on("data", (data: Buffer) => {
      const chunk = data.toString("utf-8");
      stdout += chunk;
      options?.onStdout?.(chunk);
    });

    proc.stderr?.on("data", (data: Buffer) => {
      const chunk = data.toString("utf-8");
      stderr += chunk;
      options?.onStderr?.(chunk);
    });

    if (options?.signal) {
      if (options.signal.aborted) {
        proc.kill("SIGKILL");
      } else {
        options.signal.addEventListener("abort", () => proc.kill("SIGKILL"), {
          once: true,
        });
      }
    }

    proc.on("close", (code) => {
      if (finished) return;
      finish(
        Effect.succeed({
          exitCode: code ?? 1,
          stdout,
          stderr,
          containerId: name,
        }),
      );
    });
  });

const dockerBuild: ContainerBackend["build"] = (contextDir, tag, options) =>
  Effect.async<void, ContainerError>((resume) => {
    const args: string[] = ["build", "-t", tag];
    if (options?.file) {
      args.push("-f", options.file);
    }
    if (options?.memoryLimit) {
      args.push("--memory", options.memoryLimit);
    }
    if (options?.cpuLimit) {
      args.push("--cpus", String(options.cpuLimit));
    }
    args.push(contextDir);

    const proc = spawn("docker", args, { stdio: "ignore" });

    const finish = (result: Effect.Effect<void, ContainerError>) => resume(result);

    proc.on("error", (error) =>
      finish(
        Effect.fail(
          new ContainerError(
            "start_failed",
            `Failed to start docker build: ${error.message}`,
          ),
        ),
      ),
    );

    proc.on("close", (code) => {
      if ((code ?? 1) !== 0) {
        finish(
          Effect.fail(
            new ContainerError(
              "execution_failed",
              `docker build exited with code ${code ?? 1}`,
              code ?? undefined,
            ),
          ),
        );
        return;
      }
      finish(Effect.succeed(undefined));
    });
  });

const dockerBackend: ContainerBackend = {
  name: "docker",
  isAvailable: () =>
    Effect.try(() => hasDockerCommand()).pipe(Effect.catchAll(() => Effect.succeed(false))),
  run: dockerRun,
  build: dockerBuild,
};

export const dockerBackendLive = Layer.succeed(ContainerBackendTag, dockerBackend);
export const dockerBackendLayer = dockerBackendLive;

File: crates/sandbox/src/docker.rs
//! Docker container backend

use crate::{
    backend::{BuildOptions, ContainerBackend},
    config::{ContainerConfig, ContainerRunResult},
    error::{ContainerError, ContainerResult},
};
use async_trait::async_trait;
use std::path::Path;
use std::process::Stdio;
use std::time::Duration;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use uuid::Uuid;

const DEFAULT_TIMEOUT_MS: u64 = 120_000;

/// Docker container backend
pub struct DockerBackend;

impl DockerBackend {
    /// Create a new Docker backend
    pub fn new() -> Self {
        Self
    }

    /// Check if docker command is available
    async fn has_docker_command() -> bool {
        // Check environment override
        if std::env::var("OPENAGENTS_DOCKER_AVAILABLE").as_deref() == Ok("0") {
            return false;
        }
        if std::env::var("OPENAGENTS_DOCKER_AVAILABLE").as_deref() == Ok("1") {
            return true;
        }

        // Try to run docker --version
        match Command::new("docker")
            .arg("--version")
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
            .await
        {
            Ok(status) => status.success(),
            Err(_) => false,
        }
    }

    /// Build docker run arguments
    fn build_docker_args(
        command: &[String],
        config: &ContainerConfig,
        container_name: &str,
    ) -> Vec<String> {
        let mut args = vec![
            "run".to_string(),
            "-i".to_string(),
            "--name".to_string(),
            container_name.to_string(),
        ];

        // Auto-remove
        if config.auto_remove {
            args.push("--rm".to_string());
        }

        // Workspace volume
        args.push("-v".to_string());
        args.push(format!("{}:/workspace", config.workspace_dir.display()));

        // Additional volume mounts
        for mount in &config.volume_mounts {
            args.push("-v".to_string());
            args.push(mount.clone());
        }

        // Working directory
        args.push("-w".to_string());
        args.push(config.workdir.clone().unwrap_or_else(|| "/workspace".to_string()));

        // Memory limit
        if let Some(ref limit) = config.memory_limit {
            args.push("--memory".to_string());
            args.push(limit.clone());
        }

        // CPU limit
        if let Some(cpus) = config.cpu_limit {
            args.push("--cpus".to_string());
            args.push(cpus.to_string());
        }

        // Environment variables
        for (key, value) in &config.env {
            args.push("-e".to_string());
            args.push(format!("{}={}", key, value));
        }

        // Image
        args.push(config.image.clone());

        // Command
        args.extend(command.iter().cloned());

        args
    }
}

impl Default for DockerBackend {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl ContainerBackend for DockerBackend {
    fn name(&self) -> &'static str {
        "docker"
    }

    async fn is_available(&self) -> bool {
        Self::has_docker_command().await
    }

    async fn run(
        &self,
        command: &[String],
        config: &ContainerConfig,
    ) -> ContainerResult<ContainerRunResult> {
        // Verify workspace exists
        if !config.workspace_dir.exists() {
            return Err(ContainerError::start_failed(format!(
                "workspaceDir does not exist: {}",
                config.workspace_dir.display()
            )));
        }

        let container_name = format!("oa-sbx-{}", Uuid::new_v4());
        let args = Self::build_docker_args(command, config, &container_name);

        // Spawn docker process
        let mut child = Command::new("docker")
            .args(&args)
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|e| ContainerError::start_failed(format!("Failed to start docker: {}", e)))?;

        let stdout_handle = child.stdout.take();
        let stderr_handle = child.stderr.take();

        // Read stdout
        let stdout_task = tokio::spawn(async move {
            let mut output = String::new();
            if let Some(stdout) = stdout_handle {
                let mut reader = BufReader::new(stdout);
                let mut line = String::new();
                while reader.read_line(&mut line).await.unwrap_or(0) > 0 {
                    output.push_str(&line);
                    line.clear();
                }
            }
            output
        });

        // Read stderr
        let stderr_task = tokio::spawn(async move {
            let mut output = String::new();
            if let Some(stderr) = stderr_handle {
                let mut reader = BufReader::new(stderr);
                let mut line = String::new();
                while reader.read_line(&mut line).await.unwrap_or(0) > 0 {
                    output.push_str(&line);
                    line.clear();
                }
            }
            output
        });

        // Wait with timeout
        let timeout_ms = config
            .timeout
            .map(|d| d.as_millis() as u64)
            .unwrap_or(DEFAULT_TIMEOUT_MS);
        let timeout = Duration::from_millis(timeout_ms);

        let result = tokio::time::timeout(timeout, child.wait()).await;

        match result {
            Ok(Ok(status)) => {
                let stdout = stdout_task.await.unwrap_or_default();
                let stderr = stderr_task.await.unwrap_or_default();

                Ok(ContainerRunResult {
                    exit_code: status.code().unwrap_or(1),
                    stdout,
                    stderr,
                    container_id: Some(container_name),
                })
            }
            Ok(Err(e)) => Err(ContainerError::execution_failed(
                format!("Docker process error: {}", e),
                None,
            )),
            Err(_) => {
                // Timeout - kill the container
                let _ = Command::new("docker")
                    .args(["kill", &container_name])
                    .output()
                    .await;

                Err(ContainerError::timeout(format!(
                    "Container execution timed out after {}ms",
                    timeout_ms
                )))
            }
        }
    }

    async fn build(
        &self,
        context_dir: &Path,
        tag: &str,
        options: Option<BuildOptions>,
    ) -> ContainerResult<()> {
        let mut args = vec!["build".to_string(), "-t".to_string(), tag.to_string()];

        if let Some(ref opts) = options {
            if let Some(ref file) = opts.file {
                args.push("-f".to_string());
                args.push(file.clone());
            }
            if let Some(ref limit) = opts.memory_limit {
                args.push("--memory".to_string());
                args.push(limit.clone());
            }
            if let Some(cpus) = opts.cpu_limit {
                args.push("--cpus".to_string());
                args.push(cpus.to_string());
            }
        }

        args.push(context_dir.display().to_string());

        let output = Command::new("docker")
            .args(&args)
            .output()
            .await
            .map_err(|e| ContainerError::start_failed(format!("Failed to start docker build: {}", e)))?;

        if !output.status.success() {
            return Err(ContainerError::execution_failed(
                format!(
                    "docker build exited with code {}: {}",
                    output.status.code().unwrap_or(1),
                    String::from_utf8_lossy(&output.stderr)
                ),
                output.status.code(),
            ));
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_docker_args() {
        let config = ContainerConfig::new("ubuntu:latest", "/tmp/workspace")
            .workdir("/app")
            .memory_limit("4G")
            .cpu_limit(2.0)
            .env("FOO", "bar")
            .volume_mount("/tmp/creds:/root/.claude:ro");

        let args = DockerBackend::build_docker_args(
            &["bash".to_string(), "-c".to_string(), "echo hello".to_string()],
            &config,
            "test-container",
        );

        assert!(args.contains(&"run".to_string()));
        assert!(args.contains(&"--rm".to_string()));
        assert!(args.contains(&"-v".to_string()));
        assert!(args.contains(&"--memory".to_string()));
        assert!(args.contains(&"4G".to_string()));
        assert!(args.contains(&"--cpus".to_string()));
        assert!(args.contains(&"2".to_string()));
        assert!(args.contains(&"FOO=bar".to_string()));
        assert!(args.contains(&"ubuntu:latest".to_string()));
    }
}
